subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -2350,30 +2374,59 @@ size_t lwm2m_engine_get_opaque_more(struct lwm2m_input_context *in,
 static int lwm2m_write_handler_opaque(struct lwm2m_engine_obj_inst *obj_inst,
 				      struct lwm2m_engine_res *res,
 				      struct lwm2m_engine_res_inst *res_inst,
-				      struct lwm2m_input_context *in,
+				      struct lwm2m_message *msg,
 				      void *data_ptr, size_t data_len)
 {
 	size_t len = 1;
 	bool last_pkt_block = false;
 	int ret = 0;
 	bool last_block = true;
 	struct lwm2m_opaque_context opaque_ctx = { 0 };
+	void *write_buf;
+	size_t write_buf_len;
 
-	if (in->block_ctx != NULL) {
-		last_block = in->block_ctx->last_block;
+	if (msg->in.block_ctx != NULL) {
+		last_block = msg->in.block_ctx->last_block;
 
 		/* Restore the opaque context from the block context, if used.
 		*/
-		opaque_ctx = in->block_ctx->opaque;
+		opaque_ctx = msg->in.block_ctx->opaque;
+	}
+
+	/* In case validation callback is present, write data to the temporary
+	 * buffer first, for validation. Otherwise, write to the data buffer
+	 * directly.
+	 */
+	if (res->validate_cb) {
+		write_buf = msg->ctx->validate_buf;
+		write_buf_len = sizeof(msg->ctx->validate_buf);
+	} else {
+		write_buf = data_ptr;
+		write_buf_len = data_len;
 	}
 
 	while (!last_pkt_block && len > 0) {
-		len = engine_get_opaque(in, (uint8_t *)data_ptr,
-					data_len, &opaque_ctx, &last_pkt_block);
+		len = engine_get_opaque(&msg->in, write_buf,
+					MIN(data_len, write_buf_len),
+					&opaque_ctx, &last_pkt_block);
 		if (len == 0) {
 			/* ignore empty content and continue */
 			return 0;
 		}
+
+		if (res->validate_cb) {
+			ret = res->validate_cb(
+				obj_inst->obj_inst_id, res->res_id,
+				res_inst->res_inst_id, write_buf, len,
+				last_pkt_block && last_block, opaque_ctx.len);
+			if (ret < 0) {
+				/* Bad Request. */
+				return -EEXIST;
```

Comment:
Why such an error code, if above -EINVAL is used?

Reply:
`lwm2m_engine_set()` where I return -EINVAL is a function that can be invoked from the application to set a resource data, hence it makes sense to return -EINVAL back to the application which suggests the data provided was not valid.

Here, I validate the data that comes from the network, and the engine generates the response based on the error code here: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/lwm2m/lwm2m_engine.c#L4000
Not very intuitive, but in order to gerate BAD REQUEST response, `-EEXIST` error code is expected.